pkg/storage/unified/resource/storage_backend.go

```diff
@@ -364,64 +364,57 @@ func (k *kvStorageBackend) ListIterator(ctx context.Context, req *resourcepb.Lis
 			break
 		}
 	}
+	// Create pull-style iterator from BatchGet
+	next, stop := iter.Pull2(k.dataStore.BatchGet(ctx, keys))
+
 	iter := kvListIterator{
-		keys:         keys,
-		currentIndex: -1,
-		ctx:          ctx,
-		listRV:       listRV,
-		offset:       offset,
-		limit:        req.Limit + 1, // TODO: for now we need at least one more item. Fix the caller
-		dataStore:    k.dataStore,
+		listRV: listRV,
+		offset: offset,
+		next:   next,
+		stop:   stop,
 	}
 	err := cb(&iter)
 	if err != nil {
+		iter.stop() // Clean up the iterator
 		return 0, err
 	}
+	iter.stop() // Clean up the iterator
 
 	return listRV, nil
 }
 
 // kvListIterator implements ListIterator for KV storage
 type kvListIterator struct {
-	ctx          context.Context
-	keys         []DataKey
-	currentIndex int
-	dataStore    *dataStore
-	listRV       int64
-	offset       int64
-	limit        int64
-
-	// current
-	rv    int64
-	err   error
-	value []byte
-}
+	listRV int64
+	offset int64
 
-func (i *kvListIterator) Next() bool {
-	i.currentIndex++
+	// pull-style iterator
+	next func() (DataObj, error, bool)
+	stop func()
 
-	if i.currentIndex >= len(i.keys) {
-		return false
-	}
+	// current item state
+	currentDataObj *DataObj
+	value          []byte
+}
 
-	if int64(i.currentIndex) >= i.limit {
+func (i *kvListIterator) Next() bool {
+	// Pull next item from the iterator
+	dataObj, err, ok := i.next()
+	if !ok {
 		return false
 	}
-
-	i.rv, i.err = i.keys[i.currentIndex].ResourceVersion, nil
-
-	data, err := i.dataStore.Get(i.ctx, i.keys[i.currentIndex])
 	if err != nil {
```

Comment:
Shall we store this error into iterator and return from `Error()` call?

Reply:
Also good call!